### PR TITLE
SEQNG-846: Show system activity until full completion

### PR DESCRIPTION
**Pull request by sraaphorst**
This is a fix for the issue where when configuring a subsystem (TCS, gcal, instrument), the spinning arrow button indicating activity is displayed, but then immediately vanishes.

### CODE REVIEW

File: modules/giapi/src/main/scala/giapi/client/package.scala [hunk 1]
```diff
@@ -292,6 +292,7 @@ package client {
             IO.sleep(timeout) *>
             IO.pure(CommandResult(Response.COMPLETED))
           } else {
+            IO.sleep(20.seconds) *>
```
Comment (sraaphorst) [inline modules/giapi/src/main/scala/giapi/client/package.scala]: This will have to be removed, obviously.

Comment (cquiroz) [reply modules/giapi/src/main/scala/giapi/client/package.scala]: Not sure, we may keep it shorter, like 5 secs to make the simulator more realistic

File: modules/seqexec/engine/src/main/scala/seqexec/engine/Engine.scala [hunk 2]
```diff
@@ -64,7 +65,7 @@ class Engine[D, U](stateL: Engine.State[D]) {
   def reload(seq: Sequence.State[IO], steps: List[Step[IO]]): Sequence.State[IO] =
     Sequence.State.reload(steps, seq)
 
-  def startSingle(c: ActionCoords): HandleType[EventResult.Outcome] = get.flatMap { st =>
+  def startSingle(c: ActionCoords, sys: Resource): HandleType[EventResult.Outcome] = get.flatMap { st =>
```
Comment (cquiroz) [inline modules/seqexec/engine/src/main/scala/seqexec/engine/Engine.scala]: isn't there a `Resource` in `ActionCoords`? If not that looks the right place to have it. @jluhrs WDYT?

File: modules/seqexec/model/src/main/scala/seqexec/model/enum/SingleActionOp.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+object SingleActionOp {
+  final case class Started(sid: Observation.Id, resource: Resource)
+    extends SingleActionOp
+  final case class Completed(sid: Observation.Id, resource: Resource)
+    extends SingleActionOp
+  final case class Error(sid: Observation.Id, resource: Resource)
+    extends SingleActionOp
+
+  implicit val equal: Eq[SingleActionOp] = Eq.instance {
```
Comment (cquiroz) [inline modules/seqexec/model/src/main/scala/seqexec/model/enum/SingleActionOp.scala]: do we have a test of the `Eq` laws?

File: modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala [hunk 4]
```diff
@@ -826,7 +827,7 @@ object SeqexecEngine extends SeqexecConfiguration {
   private val refreshSequences: Endo[EngineState] = (st:EngineState) => {
     st.sequences.map{ case (id, obsseq) => updateSequenceEndo(id, obsseq) }.foldLeft(st){case (s, f) => f(s)}
   }
-
+  private val logger = Logger.getLogger(this.getClass.getName)
```
Comment (cquiroz) [inline modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala]: Do we want to keep this?

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/components/sequence/steps/SubsystemControlCell.scala [hunk 5]
```diff
@@ -42,9 +47,11 @@ object SubsystemControlCell {
     id:     Observation.Id,
     stepId: StepId,
     r:      Resource
-  )(e:      ReactEvent): Callback =
+  )(e:      ReactEvent): Callback = {
+    if (r === Instrument.Gpi) logger.info(s"*** SubsystemControlCell requestResourceCall: $r")
```
Comment (cquiroz) [inline modules/seqexec/web/client/src/main/scala/seqexec/web/client/components/sequence/steps/SubsystemControlCell.scala]: If you need this you can do `Callback.log`

File: modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala [hunk 6]
```diff
@@ -886,8 +889,10 @@ object SeqexecEngine extends SeqexecConfiguration {
         svs)
       case engine.BreakpointReached(id)                                    => SequencePaused(id,
         svs)
-      case engine.SingleRunCompleted(_, _)                                 => NullEvent
-      case engine.SingleRunFailed(_, _)                                    => NullEvent
+      case engine.SingleRunCompleted(ActionCoords(sid, _), res, _)         =>
+        SingleActionEvent(SingleActionOp.Completed(sid, res))
```
Comment (sraaphorst) [inline modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala]: But then this (which I just fixed and tested) still doesn't trigger the event sequence I need.

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/components/sequence/steps/SubsystemControlCell.scala [hunk 7]
```diff
@@ -3,6 +3,8 @@
```
Comment (sraaphorst) [inline modules/seqexec/web/client/src/main/scala/seqexec/web/client/components/sequence/steps/SubsystemControlCell.scala]: Note all the changes in this file were just to help me track / understand, so you can ignore as they'll all be removed.

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/handlers/OperationsStateHandler.scala [hunk 8]
```diff
@@ -46,15 +52,27 @@ class OperationsStateHandler[M](modelRW: ModelRW[M, SequencesOnDisplay])
...
+//      updated(value.markOperations(id, TabOperations.resourceRun(r).set(none)))
+  }
+
+  def singleRunCompletedMessage: PartialFunction[Any, ActionResult[M]] = {
+    case ServerMessage(SingleActionEvent(SingleActionOp.Completed(id, r))) =>
+      logger.info(s"*** OperationsStateHandler singleRunCompletedMessage: r=$r")
+//      val v = value.markOperations(id, TabOperations.resourceRun(r).set(none))
+//      val f = Future(v)
+//      val e = Effect(f)
+    updated(value.markOperations(id, TabOperations.resourceRun(r).set(none))
```
Comment (sraaphorst) [inline modules/seqexec/web/client/src/main/scala/seqexec/web/client/handlers/OperationsStateHandler.scala]: This is what I would like to trigger when the configuration completes.

I'm not sure if this is the correct way to update the button (I discussed alternatives with @cquiroz but had trouble getting them to work due to missing information - e.g. step number - or implicits) so I used the same technique as in the commented out `RunResource` case, which did set the button back to the non-running state.

Comment (cquiroz) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/handlers/OperationsStateHandler.scala]: Right, the issue is that each `ActionHandler` is processin some of the events. In the case of events from the server they are handled in `ServerMessagehandler`. So `singleRuncompleted` is never called

**Comment by sraaphorst** [comment]
I am getting:

`Action: ServerMessage(SingleActionEvent(Completed(Id(Nonstandard(Some(GS),Some(Semester(2019A)),Some(Q),001),Index(1)),Gpi)))`

logged in the client when the configuration completes, but this doesn't seem to be matching `OperationsStateHandler.singleRunCompletedMessage`.

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/handlers/OperationsStateHandler.scala (hunk 8 above)
Comment (cquiroz) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/handlers/OperationsStateHandler.scala]: This is somewhat explained at
https://diode.suzaku.io/usage/Handlers.html

Comment (sraaphorst) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/handlers/OperationsStateHandler.scala]: Hmmm... I know you recommended putting this in `ServerMessageHandler`, but I don't think I have the step number information available here, which I seem to need, I think?

I'll try playing around with calls in there as per our earlier discussion, but I'm not sure how to get the missing information, and had some issue making the call with a missing implicit, which seemed weird.

Comment (sraaphorst) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/handlers/OperationsStateHandler.scala]: Yeah, checking now, and I don't have the step number available here, I don't think.

Comment (sraaphorst) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/handlers/OperationsStateHandler.scala]: Should I try seeing if I can pass it with `SingleActionOp`? That doesn't seem like it necessarily makes sense.

File: modules/seqexec/engine/src/main/scala/seqexec/engine/Engine.scala (hunk 2 above)
Comment (sraaphorst) [reply modules/seqexec/engine/src/main/scala/seqexec/engine/Engine.scala]: I don't think so:

`final case class ActionCoords(sid: Observation.Id, actCoords: ActionCoordsInSeq)`

`final case class ActionCoordsInSeq(stepId: StepId, execIdx: ExecutionIndex,
                                   actIdx: ActionIndex)`

File: modules/seqexec/model/src/main/scala/seqexec/model/enum/SingleActionOp.scala (hunk 3 above)
Comment (sraaphorst) [reply modules/seqexec/model/src/main/scala/seqexec/model/enum/SingleActionOp.scala]: I'll add that. I figured I was probably missing some law testing.
Do you know where is it? I can't seem to find it offhand.

File: modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala (hunk 4 above)
Comment (sraaphorst) [reply modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala]: No... I added a bunch of loggers and will get rid of them all, along with the log messages. This was just to help me see the flow of the code and see what code was being executed.

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/components/sequence/steps/SubsystemControlCell.scala (hunk 5 above)
Comment (sraaphorst) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/components/sequence/steps/SubsystemControlCell.scala]: 👍 Thanks!

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/handlers/OperationsStateHandler.scala (hunk 8 above)
Comment (cquiroz) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/handlers/OperationsStateHandler.scala]: Yes

Comment (sraaphorst) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/handlers/OperationsStateHandler.scala]: Never mind... I didn't refresh and see your comments. Okay, I'll see if I can add these to the event.

File: modules/seqexec/model/src/main/scala/seqexec/model/enum/SingleActionOp.scala (hunk 3 above)
Comment (cquiroz) [reply modules/seqexec/model/src/main/scala/seqexec/model/enum/SingleActionOp.scala]: `ModelSpec`

File: modules/seqexec/engine/src/main/scala/seqexec/engine/Engine.scala (hunk 2 above)
Comment (sraaphorst) [reply modules/seqexec/engine/src/main/scala/seqexec/engine/Engine.scala]: I moved it to `ActionCoordsInSeq`.

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/handlers/ServerMessagesHandler.scala [hunk 9]
```diff
@@ -171,6 +172,11 @@ class ServerMessagesHandler[M](modelRW: ModelRW[M, WebSocketsFocus])
       updated(value.copy(sequences = filterSequences(s.view)))
   }
 
+  val singleRunCompleteMessage: PartialFunction[Any, ActionResult[M]] = {
+    case ServerMessage(SingleActionEvent(SingleActionOp.Completed(sid, stepId, r))) =>
+      effectOnly(Effect(Future(RunResourceComplete(sid, stepId, r))))
```
Comment (cquiroz) [inline modules/seqexec/web/client/src/main/scala/seqexec/web/client/handlers/ServerMessagesHandler.scala]: 👍

**Comment by cquiroz** [comment]
Looks very good thanks. this was a difficult task (but still manageable as we see)

You can merge on the tetsts passing

**Comment by sraaphorst** [comment]
@jluhrs Thanks for pointing this out. I'll try to make that change.

**Comment by sraaphorst** [comment]
@jluhrs Wait... I'm looking at `SequenceGen.resourceAtCoords` now and I'm a bit confused. Which `resourceAtCoords` are you talking about?

If I'm using the `final case class SequenceGen`, I need the `Instrument` to instantiate it, which is the `Resource` in most cases. Also, I'd need the `steps`, and I don't know that I have that information available.

If I'm using the `final case class StepActionsGen`, I need far more information than I have at this point.

Could you give me an example of how to go from `ActionCoordsInSeq` or `ActionCoords` to `Resource` using this method?

**Comment by sraaphorst** [comment]
@jluhrs I made the requested changes. Thanks for the help with this!

File: modules/seqexec/engine/src/test/scala/seqexec/engine/SequenceSpec.scala [hunk 10]
```diff
@@ -198,7 +199,7 @@ class SequenceSpec extends FlatSpec {
         )
       )
 
-    val c = ActionCoordsInSeq(1, ExecutionIndex(0), ActionIndex(1))
+    val c = ActionCoordsInSeq(1, GmosS, ExecutionIndex(0), ActionIndex(1))
```
Comment (jluhrs) [inline modules/seqexec/engine/src/test/scala/seqexec/engine/SequenceSpec.scala]: You forgot to revert this change

File: modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala [hunk 11]
```diff
@@ -847,10 +847,11 @@ object SeqexecEngine extends SeqexecConfiguration {
     case UpdateQueueClear(qid)              => QueueUpdated(QueueManipulationOp.Clear(qid), svs)
     case StartQueue(qid, _)                 => QueueUpdated(QueueManipulationOp.Started(qid), svs)
     case StopQueue(qid, _)                  => QueueUpdated(QueueManipulationOp.Stopped(qid), svs)
-    case StartSysConfig(_, _, _)            => NullEvent
+    case StartSysConfig(sid, stepId, res)   => SingleActionEvent(SingleActionOp.Started(sid, stepId, res))
   }
 
-  def toSeqexecEvent(ev: executeEngine.ResultType)(svs: => SequencesQueue[SequenceView]): SeqexecEvent = ev match {
+  def toSeqexecEvent(ev: executeEngine.ResultType, qState: EngineState)
+                    (svs: => SequencesQueue[SequenceView]): SeqexecEvent = ev match {
```
Comment (jluhrs) [inline modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala]: `svs` could be removed and built from `qState`. The caller does that, and only builds it for calling `toSeqexcEvent`.

File: modules/seqexec/engine/src/test/scala/seqexec/engine/SequenceSpec.scala (hunk 10 above)
Comment (sraaphorst) [reply modules/seqexec/engine/src/test/scala/seqexec/engine/SequenceSpec.scala]: Yeah, I didn't catch that with the `seqexec_server/test` command, which is what I have been using for testing. I think I need to execute a more general test command since things are slipping by.

That should be fixed now.

File: modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala (hunk 11 above)
Comment (sraaphorst) [reply modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala]: Sure. I can do that, but some of the methods used are `private`, so I think I'll have to change that so I can access them. Whichever you prefer is fine with me.
